orc/src/main/java/org/apache/iceberg/orc/OrcIterable.java

```diff
@@ -73,7 +78,15 @@
   public CloseableIterator<T> iterator() {
     Reader orcFileReader = ORC.newFileReader(file, config);
     addCloseable(orcFileReader);
-    TypeDescription readOrcSchema = ORCSchemaUtil.buildOrcProjection(schema, orcFileReader.getSchema());
+
+    TypeDescription fileSchema = orcFileReader.getSchema();
+    final TypeDescription readOrcSchema;
+    if (ORCSchemaUtil.hasIds(fileSchema)) {
+      readOrcSchema = ORCSchemaUtil.buildOrcProjection(schema, fileSchema);
+    } else {
+      TypeDescription typeWithIds = ORCSchemaUtil.applyNameMapping(fileSchema, nameMapping);
```

Comment:
IIRC, nameMapping may not have all the fields which maybe present in the file schema. I think we need a way to prune columns which does not have an associated mapping defined. I can't say whether `buildOrcProjection` is handling that case or not.

Reply:
Seems like `buildOrcProjection` is handling it by traversing iceberg read schema to build the read projection

Reply:
Yes, `buildOrcProjection` has the logic to skip columns without IDs, that was implemented in https://github.com/apache/iceberg/pull/1140.